Run CI tests in latest ubuntu image using SPC

cevich opened the pull request:
Signed-off-by: Chris Evich <cevich@redhat.com>

cevich commented:
@rhatdan @lsm5 PTAL...I think I'm missing a few ``$DEBS`` in ``.travis.sh`` but I haven't got a clue.  Only know enough to be dangerous :grinning:

cevich commented:
**note: ** nevermind that travis thinks this is now a ruby project...I'm working on that :grinning:

cevich commented:
@rhatdan @lsm5 Error I'm seeing is:

```
go version go1.10 linux/amd64
+ make install.tools
go get -u -tags " libdm_no_deferred_remove "  github.com/cpuguy83/go-md2man
go get -u -tags " libdm_no_deferred_remove "  github.com/vbatts/git-validation
go get -u -tags " libdm_no_deferred_remove "  gopkg.in/alecthomas/gometalinter.v1
gometalinter.v1 -i
make: gometalinter.v1: Command not found
Makefile:79: recipe for target 'install.tools' failed
make: *** [install.tools] Error 127
```

It's likely some change I made / missing container env. expectation.

lsm5 commented:
seems like we need this tool https://github.com/alecthomas/gometalinter

lsm5 commented:
can't find it on ubuntu or fedora, yet

lsm5 commented:
maybe you can just `go get` it and be done with it?

cevich commented:
> maybe you can just go get it and be done with it?

See how little I know :grinning: 

okay, I'll try that...

cevich commented:
@lsm5 thx for your help.  Looks like most of the problems I'm hitting have to do with setting up the SPC to match cooked-in build needs previously provided by travis-magic.

cevich commented:
@rhatdan @TomSweeneyRedHat  okay, it took me all day, but I know from experience it's important to get this stuff correct.  **Bonus:** You can locally run the same actions as travis with ``./.run_ci_tests.sh``, including dropping into a shell if you need to monkey around.  See included comments for details.

So, it's ----> <---- close to passing, it's some simple goof I'm sure, you'll probably spot right away.  Once it's working here, the same pattern can be applied to other repos.  Failure details: https://travis-ci.org/containers/storage/jobs/347018751

```
Setting up for go 1.8
go version go1.8 linux/amd64
Build Environment:
...cut...
PATH=/repo_copy/.gimme/versions/go1.8.linux.amd64/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/repo_copy//bin
PWD=/repo_copy
GIMME_ENV=/repo_copy/.gimme/envs/go1.8.env
HOME=/repo_copy
GOROOT=/repo_copy/.gimme/versions/go1.8.linux.amd64
GOPATH=/repo_copy/
GO_VERSION=1.8
...cut...
go build -compiler gc -tags " libdm_no_deferred_remove "  ./cmd/containers-storage
cmd/containers-storage/container.go:9:2: cannot find package "github.com/containers/storage" in any of:
	/repo_copy/.gimme/versions/go1.8.linux.amd64/src/github.com/containers/storage (from $GOROOT)
	/repo_copy/src/github.com/containers/storage (from $GOPATH)
cmd/containers-storage/create.go:11:2: cannot find package "github.com/containers/storage/opts" in any of:
	/repo_copy/.gimme/versions/go1.8.linux.amd64/src/github.com/containers/storage/opts (from $GOROOT)
	/repo_copy/src/github.com/containers/storage/opts (from $GOPATH)
cmd/containers-storage/diff.go:10:2: cannot find package "github.com/containers/storage/pkg/archive" in any of:
	/repo_copy/.gimme/versions/go1.8.linux.amd64/src/github.com/containers/storage/pkg/archive (from $GOROOT)
	/repo_copy/src/github.com/containers/storage/pkg/archive (from $GOPATH)
cmd/containers-storage/container.go:10:2: cannot find package "github.com/containers/storage/pkg/mflag" in any of:
	/repo_copy/.gimme/versions/go1.8.linux.amd64/src/github.com/containers/storage/pkg/mflag (from $GOROOT)
	/repo_copy/src/github.com/containers/storage/pkg/mflag (from $GOPATH)
cmd/containers-storage/main.go:10:2: cannot find package "github.com/containers/storage/pkg/reexec" in any of:
	/repo_copy/.gimme/versions/go1.8.linux.amd64/src/github.com/containers/storage/pkg/reexec (from $GOROOT)
	/repo_copy/src/github.com/containers/storage/pkg/reexec (from $GOPATH)
cmd/containers-storage/create.go:13:2: cannot find package "github.com/opencontainers/go-digest" in any of:
	/repo_copy/.gimme/versions/go1.8.linux.amd64/src/github.com/opencontainers/go-digest (from $GOROOT)
	/repo_copy/src/github.com/opencontainers/go-digest (from $GOPATH)
Makefile:22: recipe for target 'containers-storage' failed
make: *** [containers-storage] Error 1
```

rhatdan commented on .run_ci_tests.sh:
```diff
@@ -0,0 +1,18 @@
+SPCCMD="${SPCCMD:-./.spc_ci_test.sh}"
+FQIN="docker.io/cevich/travis_ubuntu:latest"
+SPC_ARGS="--interactive --rm --privileged --ipc=host --pid=host --net=host"
+VOL_ARGS="-v ${REPO_VOL:-$PWD}:${REPO_VOL:-$PWD}:z
```
No need for :z, since  you are running in a an SPC.

cevich replied:
Oh really?  Didn't know that, thanks!   Fixed.

rhatdan commented on .spc_ci_test.sh:
```diff
@@ -0,0 +1,50 @@
+cd "$HOME"
+
+# FIXME: All a total guess, someone intelligent should fix this
+export GOPATH="$HOME/"
```
Usually `export GOPATH="$HOME/go"`

cevich replied:
Ahh, okay, I'll give that a try....

rhatdan commented on .travis.yml:
```diff
@@ -1,14 +1,40 @@
+        - GO_VERSION="stable"
+        - GO_VERSION="1.9"
+        - GO_VERSION="1.8"
+        - GO_VERSION="1.7"
```
I would drop support for 1.7.

cevich replied:
KK, it's gonekie.

rhatdan commented on .travis.yml:
```diff
@@ -1,14 +1,40 @@
+
 before_install:
     - sudo apt-get -qq update
-    - sudo apt-get -qq install btrfs-tools libdevmapper-dev
```
I think we are going to need these inside of the dockerfile?

cevich replied:
I left those in ``.spc_ci_test.sh`` as ``$DEBS``, thinking y'all wouldn't want to to touch dockerfiles.  The one for ``docker.io/cevich/travis_ubuntu:latest``, I tried to keep close to what Travis does by default  (package-wise).  The only non-packaged item is go, which gets installed before the test by that ``gimme`` tool - as travis does it.

Speaking of which...I set up ``travis_ubuntu`` to follow ``docker.io/ubuntu:latest`` automatically, which could cause surprises (if it changes in the middle of a test).  I don't see it as a likely problem though, and it can be "fixed" by just re-running the tests.  However, I'm happy to use a different arrangement or image if y'all prefer.

rhatdan commented:
@nalind PTAL

cevich commented:
> Usually export GOPATH="$HOME/go"

Nopers, same problem :sob:  I changed that, and ran ``./.run_ci_tests.sh`` but [get similar results.](https://paste.fedoraproject.org/paste/g0QKiGNCI90JWHPtfm-E4Q)

@nalind 

The ``cmd`` component of ``./cmd/containers-storage`` is significant in go, isn't it?

I ran across an issue while googling, that maybe is related to the variables ``gimme`` sets when it installs go (``$GOROOT`` for one).  IIRC, the issue was along the lines of, go searching that path first before ``$GOPATH``, and so any ``cmd/whatever`` packages(?) would be masked.  Does that ring any bells?

There's clearly something I'm not replicating properly in the SPC, that travis was doing.  Travis normally sets up this way:

$HOME == /home/travis
$PWD == /home/travis/containers/storage

I moved away from that because it sure does make a royal mess of the repo (since it's a volume mount).  I'll try tinkering some more, to get it to match travis closer.

cevich commented:
w00t!  Seems I've made some progress with the scripts and actually building something.

So, if ``$GOPATH=$HOME/go`` then the repository and CWD (for building) should both be ``$GOPATH/src/github.com/containers/storage`` but I have no idea why  :confused: 

Squashing up commits, rebasing and pushing...

cevich commented:
@rhatdan @nalind check out the latest travis results, things are building now but the test themselves are throwing some failures.  I'm able to replicate the same results locally as well:
```
[cevich@cevich tmp]$ git clone --single-branch --branch travis_use_docker https://github.com/cevich/storage.git
[cevich@cevich tmp]$ cd storage/
[cevich@cevich storage]$ ./.run_ci_tests.sh
...cut...
```
It's possible to get yourself into the environment as well with ``SPCCMD=bash ./.run_ci_tests.sh``.  From inside the SPC run ``source ./.spc_ci_test.sh`` to install go and start a build.  It'll leave you inside the SPC for additional tinkering as needed.  I'm happy to document this somewhere, given a pointer as to where you'd like it.

cevich commented:
@rhatdan @nalind  I need help, PTAL at the above comment.  I'm not sure what's needed to get the tests to pass.  Everything else CI-wise is working perfectly.  Thanks.

rhatdan commented:
I think that you might need to mount the test area as a tmpfs, otherwise you will be testing overlayfs on overlayfs which is not allowed.

rhatdan commented:
@cevich Would adding --tmpfs /tmp to your docker line fix the problems?

cevich commented:
> Would adding --tmpfs /tmp to your docker line fix the problems?

Ahh, interesting...see, I never would have thought that :grin: Let's see...

cevich commented:
@rhatdan Still no love :cry: Even tried with different tmpfs options, and ``-v /tmp:/tmp`` and ``-v /var/tmp:/tmp``...none are happy.  Esp. the sym and hard-link tests are really really mad.  Mind, I was running this on my Fedora laptop, so maybe there's some strange kernel/userspace clash w/ the ubuntu container.  I dunno, maybe this is a lost cause :disappointed:

cevich commented:
@rhatdan :scream: whoa...so the tests actually run much better on an ubuntu host than fedora, so there must be some odd interactions going on in fedora.  Looks like most of the tests are passing too, but not able to find 'bats' for some reason...

rhatdan commented:
do you need to do an apt get bats inside of the docker container?

cevich commented:
@rhatdan hmmm could be...but it wasn't needed before(?).  Is that a standard 'go' thing that maybe travis was providing? Easy enough to try adding it I s'pose...

rhatdan commented:
Bats is a test tool used in the tests directory.  It must be available in ubuntu, and it should have been required before.

rpm -q bats
bats-0.4.0-5.20160219git0360811.fc26.noarch

rhatdan commented:
Maybe it is just in the AMI, so we never had to require it.

cevich commented:
@rhatdan Thanks, kk 'standard stuff''...well look at that!  The CI jobs are passing now for Go 1.8 and 1.9.  Failing for 1.10...is that expected / want me to turn that test off?

rhatdan commented:
Ok lets merge this and we can look at golang 1.10 when we get there.
LGTM

cevich commented:
> Ok lets merge this and we can look at golang 1.10 when we get there.

Sounds good.

Is there somewhere I can add documentation on the setup, and running the tests locally?
